swap_server_client.go

```diff
@@ -275,8 +301,151 @@ func (s *grpcSwapServerClient) NewLoopInSwap(ctx context.Context,
-	s.conn.Close()
+// ServerUpdate summarizes an update from the swap server.
+type ServerUpdate struct {
+	// State is the state that the server has sent us.
+	State looprpc.ServerSwapState
+
+	// Timestamp is the time of the server state update.
+	Timestamp time.Time
+}
+
+// SubscribeLoopInUpdates subscribes to loop in server state and pipes updates
+// into the channel provided.
+func (s *grpcSwapServerClient) SubscribeLoopInUpdates(ctx context.Context,
+	hash lntypes.Hash) (<-chan *ServerUpdate, <-chan error, error) {
+
+	resp, err := s.server.SubscribeLoopInUpdates(
+		ctx, &looprpc.SubscribeUpdatesRequest{
+			ProtocolVersion: protocolVersion,
+			SwapHash:        hash[:],
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	receive := func() (*ServerUpdate, error) {
+		response, err := resp.Recv()
+		if err != nil {
+			return nil, err
+		}
+
+		return &ServerUpdate{
+			State:     response.State,
+			Timestamp: time.Unix(0, response.TimestampNs),
+		}, err
+	}
+
+	updateChan, errChan := s.makeServerUpdate(ctx, receive)
+	return updateChan, errChan, nil
+}
+
+// SubscribeLoopOutUpdates subscribes to loop out server state and pipes updates
+// into the channel provided.
+func (s *grpcSwapServerClient) SubscribeLoopOutUpdates(ctx context.Context,
+	hash lntypes.Hash) (<-chan *ServerUpdate, <-chan error, error) {
+
+	resp, err := s.server.SubscribeLoopOutUpdates(
+		ctx, &looprpc.SubscribeUpdatesRequest{
+			ProtocolVersion: protocolVersion,
+			SwapHash:        hash[:],
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	receive := func() (*ServerUpdate, error) {
+		response, err := resp.Recv()
+		if err != nil {
+			return nil, err
+		}
+
+		return &ServerUpdate{
+			State:     response.State,
+			Timestamp: time.Unix(0, response.TimestampNs),
+		}, err
+	}
+
+	updateChan, errChan := s.makeServerUpdate(ctx, receive)
+	return updateChan, errChan, nil
+}
+
+// makeServerUpdate takes a stream receive function and a channel that it
+// should pipe updates into. It sends events into the updates channel until
+// the client cancels, server client shuts down or the subscription is cancelled
+// server side.
+func (s *grpcSwapServerClient) makeServerUpdate(ctx context.Context,
+	receive func() (*ServerUpdate, error)) (<-chan *ServerUpdate,
+	<-chan error) {
+
+	// We will return exactly one error from this function so we buffer
+	// our error channel so that the function exit is not dependent on
+	// the error being read.
+	errChan := make(chan error, 1)
+	updateChan := make(chan *ServerUpdate)
+
+	// Create a goroutine that will pipe updates in to our updates channel.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		for {
+			// Try to receive from our stream. If there are no items
+			// to consume, this call will block. If our stream is
+			// cancelled by the server we will receive an error.
+			response, err := receive()
+			switch err {
+			// If we get a nil error, we fallthrough to deliver
```

Comment:
nit: `fallthrough` is perhaps the wrong word to use here as there's also a keyword that does the exact opposite of what's happening here. Perhaps "If we get a nil error, we proceed with delivering the update we have just received"?